### PR TITLE
[SYCL][ABI-Break] Drop pi::assertion

### PR DESCRIPTION
**Pull request by AlexeySachkov**
This function should not be exported and looking at its uses it seems like they call can be replaced with regular `assert` or a `static_assert` without hurting any functionality.

This is patch is essentially a by-product of intel/llvm#14145 and it is done to simplify that change, i.e. PI plugins removal should not be ABI-breaking by itself, we just need to cleanup some of our exported symbols.

### CODE REVIEW

File: sycl/include/sycl/detail/pi.hpp [hunk 1]
```diff
@@ -260,9 +258,7 @@ namespace pi {
 // Want all the needed casts be explicit, do not define conversion
 // operators.
 template <class To, class From> inline To cast(From value) {
-  // TODO: see if more sanity checks are possible.
-  sycl::detail::pi::assertion((sizeof(From) == sizeof(To)),
-                              "assert: cast failed size check");
+  static_assert(sizeof(From) == sizeof(To), "cast failed size check");
```
Comment (steffenlarsen) [inline sycl/include/sycl/detail/pi.hpp:261]: This could even have performance benefits! 🥳

**Review by steffenlarsen — APPROVED** [review]

**Review by kbenzie — APPROVED** [review]

File: sycl/include/sycl/detail/backend_traits_opencl.hpp [hunk 2]
```diff
@@ -143,10 +143,9 @@ namespace pi {
 // Cast for std::vector<cl_event>, according to the spec, make_event
 // should create one(?) event from a vector of cl_event
 template <class To> inline To cast(std::vector<cl_event> value) {
-  sycl::detail::pi::assertion(
-      value.size() == 1,
-      "Temporary workaround requires that the "
-      "size of the input vector for make_event be equal to one.");
+  assert(value.size() == 1 &&
```
Comment (aelovikov-intel) [inline sycl/include/sycl/detail/backend_traits_opencl.hpp:146]: technically, it's different behavior in release build (where new `assert` will be simply ignored), but I'm fine with that.

**Review by aelovikov-intel — APPROVED** [review]

**Comment by AlexeySachkov** [comment]
```
TIMEOUT: SYCL :: OneapiDeviceSelector/sub-devices.cpp (1969 of 2106)
******************** TEST 'SYCL :: OneapiDeviceSelector/sub-devices.cpp' FAILED ********************
Exit Code: -9
Timeout: Reached timeout of 600 seconds

Command Output (stdout):
--
# RUN: at line 9
/__w/llvm/llvm/toolchain/bin//clang++   -fsycl -fsycl-targets=nvptx64-nvidia-cuda  /__w/llvm/llvm/llvm/sycl/test-e2e/OneapiDeviceSelector/sub-devices.cpp -o /__w/llvm/llvm/build-e2e/OneapiDeviceSelector/Output/sub-devices.cpp.tmp.out
# executed command: /__w/llvm/llvm/toolchain/bin//clang++ -fsycl -fsycl-targets=nvptx64-nvidia-cuda /__w/llvm/llvm/llvm/sycl/test-e2e/OneapiDeviceSelector/sub-devices.cpp -o /__w/llvm/llvm/build-e2e/OneapiDeviceSelector/Output/sub-devices.cpp.tmp.out
# note: command had no output on stdout or stderr
# RUN: at line 14
env ONEAPI_DEVICE_SELECTOR="*:*.*" env SYCL_PI_CUDA_ENABLE_IMAGE_SUPPORT=1  /__w/llvm/llvm/build-e2e/OneapiDeviceSelector/Output/sub-devices.cpp.tmp.out
# executed command: env 'ONEAPI_DEVICE_SELECTOR=*:*.*' env SYCL_PI_CUDA_ENABLE_IMAGE_SUPPORT=1 /__w/llvm/llvm/build-e2e/OneapiDeviceSelector/Output/sub-devices.cpp.tmp.out
# .---command stdout------------
# | 0 devices found.
# `-----------------------------
# RUN: at line 15
env ONEAPI_DEVICE_SELECTOR="*:*.*.*" env SYCL_PI_CUDA_ENABLE_IMAGE_SUPPORT=1  /__w/llvm/llvm/build-e2e/OneapiDeviceSelector/Output/sub-devices.cpp.tmp.out
# executed command: env 'ONEAPI_DEVICE_SELECTOR=*:*.*.*' env SYCL_PI_CUDA_ENABLE_IMAGE_SUPPORT=1 /__w/llvm/llvm/build-e2e/OneapiDeviceSelector/Output/sub-devices.cpp.tmp.out
# .---command stdout------------
# | 0 devices found.
# `-----------------------------
# error: command failed with exit status: -9
# error: command reached timeout: True
```

I doubt that it is caused by my changes, but I don't know if that's a known issue or not

**Comment by AlexeySachkov** [comment]
I've reported the failure in #14653. Considering that this patch should effectively be a non-functional change, I will proceed with the merge.

**Pull request merged**